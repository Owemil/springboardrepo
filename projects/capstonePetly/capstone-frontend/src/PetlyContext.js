import React from "react";

const PetlyContext = React.createContext();

export default PetlyContext;