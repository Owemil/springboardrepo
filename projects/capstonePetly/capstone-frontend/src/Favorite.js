import React, { useContext, useState } from "react";
import PetlyContext from "./PetlyContext";
import { useNavigate } from "react-router-dom";

// Favorite component takes in org and pet data and if currFav is true or false
// it renders a small heart on all animal and org cards

function Favorite({ pet, org, currFav = false }) {
    const [fav, setFav] = useState(currFav)
    const navigate = useNavigate()
    const { user, addFavPet, removeFavPet, addFavOrg, removeFavOrg, setReRender } = useContext(PetlyContext)
    const emptyHeart = "\u2661"
    const filledHeart = "\u{1F497}"

    const handleClick = async (evt) => {
        evt.stopPropagation()
        if (user) {
            if (pet) {

                !fav ? await addFavPet(user.username, pet) : await removeFavPet(user.username, pet.petId)
                setFav(!fav)
            } else if (org) {

                !fav ? await addFavOrg(user.username, org) : await removeFavOrg(user.username, org.orgId)
                setFav(!fav)
                setReRender('rerender')
            }

        } else {
            navigate(`/signup`)
        }
    }

    return (
        <>
            {!fav ?
                <div className="fav-btn" onClick={handleClick}>
                    <h1 className="fav-icon">{emptyHeart}</h1>
                </div>

                :
                <div className="fav-btn" onClick={handleClick}>
                    <h1 className="fav-icon">{filledHeart}</h1>
                </div>
            }

        </>
    )
}


export default Favorite