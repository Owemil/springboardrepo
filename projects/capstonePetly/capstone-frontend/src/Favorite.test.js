import React from "react";
import PetlyContext from "./PetlyContext.js";
import { fireEvent, render, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Favorite from "./Favorite.js"

const user = { data: "I just needed something here" }
const nullUser = null
const addFavPet = jest.fn()
const removeFavPet = jest.fn()
const addFavOrg = jest.fn()
const removeFavOrg = jest.fn()
const setReRender = jest.fn()
const emptyHeart = "\u2661"
const filledHeart = "\u{1F497}"
const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

it("should render (pet, not fav)", function () {
    render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)
})
it("should render (pet, fav)", function () {
    render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)
})
it("should render (org, not fav)", function () {
    render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)
})
it("should render (org, fav)", function () {
    render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)
})

it("should match snapshot (pet, not fav)", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    expect(asFragment).toMatchSnapshot()
})
it("should match snapshot (pet, fav)", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)

    expect(asFragment).toMatchSnapshot()
})
it("should match snapshot (org, not fav)", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    expect(asFragment).toMatchSnapshot()
})
it("should match snapshot (org, fav)", function () {
    const { asFragment } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)

    expect(asFragment).toMatchSnapshot()
})

it("should navigate when no user (pet)", function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: nullUser,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(emptyHeart)

    fireEvent.click(favDiv)

    expect(mockedUsedNavigate).toHaveBeenCalled()
})
it("should navigate when no user (org)", function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: nullUser,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(emptyHeart)

    fireEvent.click(favDiv)

    expect(mockedUsedNavigate).toHaveBeenCalled()
})

it("should fav when clicked (pet)", async function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(emptyHeart)

    fireEvent.click(favDiv)
    waitFor(() => {
        expect(filledHeart).toBeInTheDocument()
        expect(addFavPet).toHaveBeenCalled()
    })
})
it("should fav when clicked (org)", async function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(emptyHeart)

    fireEvent.click(favDiv)
    waitFor(() => {
        expect(filledHeart).toBeInTheDocument()
        expect(addFavOrg).toHaveBeenCalled()
    })
})

it("should unfav when clicked (pet)", async function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite pet="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(filledHeart)

    fireEvent.click(favDiv)
    waitFor(() => {
        expect(emptyHeart).toBeInTheDocument()
        expect(removeFavPet).toHaveBeenCalled()
    })
})
it("should unfav when clicked (org)", async function () {
    const { getByText } = render(
        <MemoryRouter>
            <PetlyContext.Provider value={{
                user: user,
                addFavPet: addFavPet,
                removeFavPet: removeFavPet,
                addFavOrg: addFavOrg,
                removeFavOrg: removeFavOrg,
                setReRender: setReRender
            }}>
                <Favorite org="1" currFav={true} />
            </PetlyContext.Provider>
        </MemoryRouter>)

    const favDiv = getByText(filledHeart)

    fireEvent.click(favDiv)
    waitFor(() => {
        expect(emptyHeart).toBeInTheDocument()
        expect(removeFavOrg).toHaveBeenCalled()
    })
})