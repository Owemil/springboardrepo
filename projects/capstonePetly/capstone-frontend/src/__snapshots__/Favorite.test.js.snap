// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot (org, fav) 1`] = `[Function]`;

exports[`should match snapshot (org, not fav) 1`] = `[Function]`;

exports[`should match snapshot (pet, fav) 1`] = `[Function]`;

exports[`should match snapshot (pet, not fav) 1`] = `[Function]`;
